heltour/tournament/tasks.py

```diff
@@ -306,6 +312,95 @@ def update_slack_users():
+               # set expiration of rejected token to yesterday, so we know to not use it anymore.
+               for game in league_games:
+                   if game.white.oauth_token.access_token==bad_token:
+                       game.white.oauth_token.expires = timezone.now() + timedelta(days=-1)
+                       game.white.oauth_token.save()
+                       break # only one oauth_token can be the bad token, so we do not need to proceed the loop
+                   if game.black.oauth_token.access_token==bad_token:
+                       game.black.oauth_token.expires = timezone.now() + timedelta(days=-1)
+                       game.black.oauth_token.save()
+                       break
+               # remove bad token from our token string + the good token paired with it, remove potential superfluous comma
+               new_tokens = re.sub('^,', '', re.sub(f'(^|,)([A-z0-9_]*:)?{bad_token}(:[A-z0-9_]*)?', '', tokens))
+           if new_tokens:
+               try:
+                   # if there are still tokens to be paired, retry, and give up afterwards.
+                   result = lichessapi.bulk_start_games(tokens=new_tokens, clock=clock, increment=increment, clockstart=clockstart, variant=variant, leaguename=leaguename)
+               except lichessapi.ApiClientError as err:
+                   # give up.
+                   logger.exception(f'[ERROR] Failed to bulk start games for league {leaguename} after removing rejected tokens.')
+           else: # no tokens left after deleting rejected tokens
+               result = None
+        except KeyError:
+            logger.exception(f'[ERROR] could not parse error as json for {leaguename}:\n{e}')
+    # use lichess reply to set game ids
+    for game in league_games:
+        try:
+            for gameids in result['games']:
+                if (gameids['white'] == game.white.lichess_username.lower() and
+                   gameids['black'] == game.black.lichess_username.lower()):
+                       game.game_link = get_gamelink_from_gameid(gameids['id'])
+                       game.save()
+                       signals.notify_players_game_started.send(sender=_start_league_games,
+                                                                pairing=game,
+                                                                gameid=gameids['id'])
+        except KeyError:
+            logger.info(f'[ERROR] For league {leaguename}, unexpected bulk pairing json response with error {e}')
+        except TypeError: # if all tokens are rejected by lichess, result['games'] is None, resulting in a TypeError.
+            pass
+
+
+@app.task()
+def start_games():
+    logger.info('[START] Checking for games to start.')
+    games_to_start = PlayerPairing.objects.filter(
+            result='', game_link='',
+            scheduled_time__lt=timezone.now() + timedelta(minutes=5, seconds=30),
+            scheduled_time__gt=timezone.now() + timedelta(seconds=30),
```

Comment:
> can I just check these times are correct?

tl;dr: yes, i did break my pretty little head about those times, and i picked those on purpose. no, i do not think i necessarily picked the best choice.

> would it also not be a good idea to pick up games in the past as well?

i thought about it, but i am honestly not sure. there is a lot of weird stuff that could happen i think. for example, players play a game that is not picked up for some reason and before the mods can add the gamelink, heltour starts a new game for them. that wouldn't be the end of the world of course. but if the game is ongoing, it might distract them.
or maybe players manage to somehow schedule a game for a time that already happened by accident. i think chesster would not allow this, but mods can manually schedule games, and there's less checking there. would still not be the end of the world of course.

at the same time, i do not really see the point in automatically starting games that should have happened a while ago. if the game did not start, or heltour thinks the game did not start, there might be a reason for that, and it might be better to not start it automatically.

anyway, the logic about all time points is definitely debatable. some people might be pissed about games starting 5 minutes early, prohibiting them to play some last minute blitz. same with the question of when exactly to start the clock. currently it's 6 minutes after the game was started, which will be 1 minute after the game was scheduled for most people.

i am absolutely happy to discuss these times, i am not at all confident that i picked the best options. very glad you looked at it.

Reply:
That all sounds reasonable, and it can all be adjusted later if needed.